File: problems/0028-find-the-index-of-the-first-occurrence-in-a-string/solution.py
"""
Problem Number: 28
Problem Name: Find the Index of the First Occurrence in a String
Difficulty: Easy
Topics: Two Pointers, String, String Matching
Companies: Microsoft, Facebook, Apple
"""

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        """
        Approach: Sliding Window (or just built-in find)
        Time Complexity: O(n * m) - naive implementation
        Space Complexity: O(1)
        """
        if not needle:
            return 0
            
        return haystack.find(needle)

# Test cases
if __name__ == "__main__":
    solution = Solution()
    print(solution.strStr("sadbutsad", "sad")) # 0
    print(solution.strStr("leetcode", "leeto")) # -1

File: problems/0028-find-the-index-of-the-first-occurrence-in-a-string/solution.js
/**
 * @problemNumber: 28
 * @problemName: Find the Index of the First Occurrence in a String
 * @difficulty: Easy
 * @topics: Two Pointers, String, String Matching
 * @companies: Microsoft, Facebook, Apple
 */

/**
 * Approach: Sliding Window
 * Time Complexity: O(n * m)
 * Space Complexity: O(1)
 *
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
function strStr(haystack, needle) {
    if (needle.length === 0) return 0;

    for (let i = 0; i <= haystack.length - needle.length; i++) {
        let match = true;
        for (let j = 0; j < needle.length; j++) {
            if (haystack[i + j] !== needle[j]) {
                match = false;
                break;
            }
        }
        if (match) return i;
    }

    return -1;
}

// Test cases
console.log(strStr("sadbutsad", "sad")); // 0
console.log(strStr("leetcode", "leeto")); // -1

module.exports = { strStr };
